src/gpio.rs

```diff
@@ -293,1040 +377,841 @@ macro_rules! gpio {
-                    fn set_speed(&mut self, cr: &mut $CR, speed: IOPinSpeed){
-                        const OFFSET: u32 = (4 * $i) % 32;
+macro_rules! cr {
+    ($CR:ident, $cr:ident) => {
+        #[doc(hidden)]
+        pub struct $CR {
+            _0: (),
+        }
 
-                        cr.cr().modify(|r, w| unsafe {
-                            w.bits((r.bits() & !(0b11 << OFFSET)) | ((speed as u32) << OFFSET))
-                        });
-                    }
-                }
+        impl<const P: char> Cr<$CR, P> {
+            // NOTE(allow) we get a warning on GPIOC because it only has 3 high pins
+            #[allow(dead_code)]
+            pub(crate) fn cr(&mut self) -> &pac::gpioa::$CR {
+                unsafe { &(*Gpio::<P>::ptr()).$cr }
+            }
+        }
 
-                impl OutputSpeed<$CR> for $PXi<Alternate<PushPull>> {
-                    fn set_speed(&mut self, cr: &mut $CR, speed: IOPinSpeed){
-                        const OFFSET: u32 = (4 * $i) % 32;
+        impl<MODE, const P: char, const N: u8> Pin<MODE, $CR, P, N>
+        where
+            MODE: Active,
+        {
+            /// Configures the pin to operate as an alternate function push-pull output
+            /// pin.
+            #[inline]
+            pub fn into_alternate_push_pull(
+                self,
+                cr: &mut Cr<$CR, P>,
+            ) -> Pin<Alternate<PushPull>, $CR, P, N> {
+                // Alternate function output push pull
+                const CNF: u32 = 0b10;
+                // Output mode, max speed 50 MHz
+                const MODE: u32 = 0b11;
+                const BITS: u32 = (CNF << 2) | MODE;
+
+                // input mode
+                cr.cr().modify(|r, w| unsafe {
+                    w.bits((r.bits() & !(0b1111 << Self::OFFSET)) | (BITS << Self::OFFSET))
+                });
+
+                Pin::new(Alternate::_new())
+            }
 
-                        cr.cr().modify(|r, w| unsafe {
-                            w.bits((r.bits() & !(0b11 << OFFSET)) | ((speed as u32) << OFFSET))
-                        });
-                    }
-                }
+            /// Configures the pin to operate as an alternate function open-drain output
+            /// pin.
+            #[inline]
+            pub fn into_alternate_open_drain(
+                self,
+                cr: &mut Cr<$CR, P>,
+            ) -> Pin<Alternate<OpenDrain>, $CR, P, N> {
+                // Alternate function output open drain
+                const CNF: u32 = 0b11;
+                // Output mode, max speed 50 MHz
+                const MODE: u32 = 0b11;
+                const BITS: u32 = (CNF << 2) | MODE;
+
+                // input mode
+                cr.cr().modify(|r, w| unsafe {
+                    w.bits((r.bits() & !(0b1111 << Self::OFFSET)) | (BITS << Self::OFFSET))
+                });
+
+                Pin::new(Alternate::_new())
+            }
```

Comment:
Why can these not use the `set_mode` functions?

Reply:
Ok